### ISSUE TITLE
The memory the osim env takes increases linearly with the number of steps it took, even if the env gets resetted several times in the middle.

### ISSUE THREAD

**Issue by syllogismos** [stanfordnmbl/osim-rl#10]
The memory the python process takes increases linearly with the number of steps taken by GaitEnv, is it supposed to be like that?

```env = GaitEnv(visualize=False) 
s = 0
while s<50000:   
    o = env.reset()   
    d = False   
    while not d:   
        o, r, d, i = env.step(env.action_space.sample())   
        s += 1   
```

This small script takes more than 500MB in my computer, python 2.7 and osx

And while training, the env takes more than 20GB memory for some 1Million steps, am I doing something wrong? There are cases where my training script stopped because of memory errors after training for a day on my little 8GB computer.

**Comment by ViktorM** [comment]
I have the same problem.

**Comment by chrisdembia** [comment]
@kidzik I'd be happy to look through the code with you to identify memory leaks.

**Comment by kidzik** [comment]
Thanks I will look into that in the next few days

**Comment by kidzik** [comment]
@syllogismos @ViktorM I have troubles with reproducing the error. Can you tell me more about your environment (os, python version)? 
Can you please try this script https://github.com/stanfordnmbl/osim-rl/blob/master/tests/test.memory.py with guppy? Thanks!

**Comment by ViktorM** [comment]
OS - Ubuntu 16.04, python 2.7, your standard conda environment. I'll be able to try script not only in 2-3 days. Usual scenario for the error is just to start training of the Gait env. In between 400K and 500K iterations it gets killed. Rarely after 500K. It takes approximately more than 16 hours of training.

**Comment by alexis-jacq** [comment]
I have the same. It takes ~1h for 1000 steps, without any learning.
Is there a way to speed-up the environement's computations ? (like using GPU)

**Comment by kidzik** [comment]
I suppose you mean 1000 episodes? Or just 1000 steps?
The environment is indeed fairly difficult and time-consuming, but there are some tricks possible: you can run simulations on multiple cores and use an algorithm that can leverage parallel runs, or you can run short episodes at first.

This difficulty is a part of the problem -- finding an algorithm which can learn quickly in such a complex system is a goal on its own.

**Comment by alexis-jacq** [comment]
Yes, parallel computing is working, but I only have 4 cores. Let's move on EPFL servers...

**Comment by ctmakro** [comment]
I wrote a script which starts a RunEnv in a seperate process and communicate with it. I recreate that process every 50 episodes to limit it's memory consumption. Before that, the training process consumed ~10GB memory and crashed hard.

**Comment by AdamStelmaszczyk** [comment]
Memory leak still exists. I'm aware that one possible workaround is killing and restarting the leaking process, but maybe let's fix the leak :)

My gut feeling is that it happens during Python interaction with OpenSim's C++ code. I saw similar things happening with cap'n'proto Python API, you had to call some function in Python in particular way to prevent memleak. My guess here will be that osim-rl calls some function in OpenSim, e.g. next_frame() or something and OpenSim allocates some memory to return the result. And then it expects that some other function is called, that is never called. Or, there is simply a leak during next_frame() execution, totally inside OpenSim C++ code, but that seem to me less likely, they would probably spot it before or maybe even they auto test the code with valgrind or similar in CI.

Or maybe the Python wrappers generated with SWIG are problematic...

**Comment by kidzik** [comment]
Good point, thanks for the insight. It seems to be dependent on the environment. Can you give us the details about your environment and a script for reproducing the error?

**Comment by AdamStelmaszczyk** [comment]
Linux x86_64 4.8.0-53. Conda env as in osim-rl README, with Python 3, `tensorflow` and `psutil`:

```
conda create -n py3 -c kidzik opensim git tensorflow psutil python=3.6
pip install git+https://github.com/stanfordnmbl/osim-rl.git
```

With Python 2 it's also leaking.

Minimal working example, `memleak.py`:

```python
import os

import psutil
from osim.env.run import RunEnv
from datetime import datetime


def time():
    return datetime.now().strftime('%H:%M:%S')


def memory_used():
    process = psutil.Process(os.getpid())
    return process.memory_info().rss  # https://pythonhosted.org/psutil/#psutil.Process.memory_info


env = RunEnv(visualize=False)
env.reset()
step = 0
episode = 0
while True:
    observation, reward, done, info = env.step(env.action_space.sample())
    step += 1
    if done:
        episode += 1
        env.reset()
        print("%s Episode %s, steps %s, memory %s" % (time(), episode, step, memory_used()))
        step = 0

```

Output:
```
(py3) adam@adam-ThinkPad-T520 ~/Desktop/running/stuff $ python memleak.py
Updating Model file from 30000 to latest format...
Loaded model gait9dof18musc_Thelen_BigSpheres.osim from file /home/adam/miniconda3/envs/py3/lib/python3.5/site-packages/osim/env/../models/gait9dof18musc.osim
20:56:22 Episode 1, steps 120, memory 133820416
20:56:31 Episode 2, steps 117, memory 134230016
20:56:40 Episode 3, steps 120, memory 135041024
20:56:49 Episode 4, steps 118, memory 135041024
20:56:57 Episode 5, steps 116, memory 136663040
20:57:05 Episode 6, steps 117, memory 136663040
20:57:14 Episode 7, steps 117, memory 136663040
20:57:23 Episode 8, steps 122, memory 136663040
20:57:32 Episode 9, steps 117, memory 139636736
20:57:40 Episode 10, steps 118, memory 139636736
20:57:48 Episode 11, steps 119, memory 139636736
20:57:57 Episode 12, steps 117, memory 139636736
20:58:06 Episode 13, steps 118, memory 139636736
20:58:14 Episode 14, steps 119, memory 139636736
20:58:23 Episode 15, steps 120, memory 139636736
20:58:32 Episode 16, steps 119, memory 139636736
20:58:40 Episode 17, steps 117, memory 139636736
20:58:49 Episode 18, steps 118, memory 145854464
20:58:57 Episode 19, steps 116, memory 145854464
20:59:06 Episode 20, steps 117, memory 145854464
20:59:14 Episode 21, steps 119, memory 145854464
20:59:23 Episode 22, steps 118, memory 145854464
20:59:31 Episode 23, steps 119, memory 145854464
20:59:40 Episode 24, steps 119, memory 145854464
20:59:48 Episode 25, steps 116, memory 145854464
20:59:57 Episode 26, steps 117, memory 145854464
21:00:05 Episode 27, steps 119, memory 145854464
21:00:14 Episode 28, steps 118, memory 145854464
21:00:23 Episode 29, steps 119, memory 145854464
21:00:32 Episode 30, steps 118, memory 145854464
21:00:40 Episode 31, steps 118, memory 145854464
21:00:49 Episode 32, steps 118, memory 145854464
21:00:58 Episode 33, steps 117, memory 145854464
21:01:06 Episode 34, steps 117, memory 145854464
21:01:14 Episode 35, steps 119, memory 158560256
21:01:23 Episode 36, steps 117, memory 158560256
21:01:32 Episode 37, steps 118, memory 158560256
21:01:40 Episode 38, steps 117, memory 158560256
21:01:50 Episode 39, steps 121, memory 158560256
21:01:58 Episode 40, steps 118, memory 158560256
21:02:07 Episode 41, steps 117, memory 158560256
21:02:16 Episode 42, steps 117, memory 158560256
21:02:27 Episode 43, steps 119, memory 158560256
21:02:37 Episode 44, steps 117, memory 158560256
21:02:46 Episode 45, steps 118, memory 158560256
21:02:55 Episode 46, steps 118, memory 158560256
21:03:03 Episode 47, steps 118, memory 158560256
21:03:13 Episode 48, steps 122, memory 158560256
21:03:22 Episode 49, steps 117, memory 158560256
21:03:31 Episode 50, steps 117, memory 158560256
(...)
```
Memory usage always grew.

I noticed that if `env.step(env.action_space.sample())` line is replaced with `done = True`, then the used memory doesn't increase.

So, calls to `env.reset()` are fine, calls to `env.step()` are causing the leak.

**Comment by AdamStelmaszczyk** [comment]
If [this](https://github.com/stanfordnmbl/osim-rl/blob/master/osim/env/osim.py#L178) line is commented out, then the memory doesn't increase:

```python
manager.integrate(self.osim_model.state)
```

**Comment by AdamStelmaszczyk** [comment]
I just [searched](https://github.com/opensim-org/opensim-core/search?utf8=%E2%9C%93&q=leak&type=) for "leak" in opensim-core repo... [one search hit](https://github.com/opensim-org/opensim-core/blob/4446724f558fb15a4fb4fd6a2dbd2df3c20f9d6a/OpenSim/Common/XYFunctionInterface.cpp#L230):

```cpp
{
    double* xValues = new double[2];
    xValues[0] = -1.0;
    xValues[1] = 1.0;
    return xValues; // possible memory leak
}
```

Don't know if it's related to our problem, but this one looks like a memleak, every `new` should have a corresponding `delete` (well, or use [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization))... It's in `const double* XYFunctionInterface::getXValues() const`.

[Another](https://github.com/opensim-org/opensim-core/blob/dc17ff08e0b40311c984f978b1462f7e31e2756a/OpenSim/Utilities/simmToOpenSim/acpp/src/cccp.c#L714):

```cpp
            /* NOTE:  If we want to be ambitious here we can also determine
             *  when to free the hash node's "value" (i.e. cpval, or defn).
             *  This seemed tricky however, and I decided that a small
             *  memory leak was better than a possible invalid free().
```

**Comment by kidzik** [comment]
Thank you very very much! I will check it on my machine and discuss it with the OpenSim team.

**Comment by AdamStelmaszczyk** [comment]
@kidzik Any news?

**Comment by kidzik** [comment]
I really have troubles reproducing the error. Your code on my machine outputs
```
(opensim-rl) lukasz@lukasz-MOBILIZE-NMBL:~/workspace/osim-rl/tests$ python test.leak.py 
Updating Model file from 30000 to latest format...
Loaded model gait9dof18musc_Thelen_BigSpheres.osim from file /home/lukasz/anaconda2/envs/opensim-rl/lib/python2.7/site-packages/osim/env/../models/gait9dof18musc.osim
17:33:15 Episode 1, steps 120, memory 128196608
17:33:20 Episode 2, steps 117, memory 128196608
17:33:24 Episode 3, steps 121, memory 128196608
17:33:29 Episode 4, steps 118, memory 128196608
17:33:33 Episode 5, steps 116, memory 128196608
17:33:37 Episode 6, steps 116, memory 128196608
17:33:42 Episode 7, steps 117, memory 128196608
17:33:47 Episode 8, steps 120, memory 128196608
17:33:51 Episode 9, steps 118, memory 128196608
17:33:56 Episode 10, steps 119, memory 128196608
17:34:00 Episode 11, steps 119, memory 128196608
17:34:05 Episode 12, steps 118, memory 128196608
17:34:10 Episode 13, steps 118, memory 128196608
17:34:14 Episode 14, steps 119, memory 128196608
17:34:19 Episode 15, steps 118, memory 128196608
17:34:23 Episode 16, steps 119, memory 128196608
17:34:28 Episode 17, steps 116, memory 128196608
17:34:32 Episode 18, steps 117, memory 128196608
17:34:37 Episode 19, steps 117, memory 128196608
17:34:41 Episode 20, steps 118, memory 128196608
17:34:46 Episode 21, steps 120, memory 128196608
17:34:50 Episode 22, steps 118, memory 128196608
17:34:55 Episode 23, steps 119, memory 128196608
17:34:59 Episode 24, steps 118, memory 128196608
17:35:04 Episode 25, steps 116, memory 128196608
17:35:08 Episode 26, steps 117, memory 128196608
17:35:13 Episode 27, steps 121, memory 128196608
17:35:17 Episode 28, steps 118, memory 128196608
17:35:22 Episode 29, steps 119, memory 128196608
17:35:26 Episode 30, steps 119, memory 128196608
17:35:31 Episode 31, steps 119, memory 128196608
17:35:36 Episode 32, steps 118, memory 128196608
17:35:40 Episode 33, steps 117, memory 128196608
17:35:44 Episode 34, steps 117, memory 128196608
17:35:49 Episode 35, steps 119, memory 128196608
17:35:54 Episode 36, steps 119, memory 128196608
17:35:58 Episode 37, steps 118, memory 128196608
17:36:03 Episode 38, steps 118, memory 128196608
17:36:07 Episode 39, steps 118, memory 128196608
17:36:12 Episode 40, steps 120, memory 128196608
17:36:16 Episode 41, steps 117, memory 128196608
17:36:21 Episode 42, steps 116, memory 128196608
17:36:25 Episode 43, steps 118, memory 128196608
17:36:30 Episode 44, steps 117, memory 128196608
17:36:34 Episode 45, steps 118, memory 128196608
17:36:39 Episode 46, steps 119, memory 128196608
17:36:43 Episode 47, steps 117, memory 128196608
17:36:48 Episode 48, steps 119, memory 128196608
17:36:52 Episode 49, steps 118, memory 128196608
17:36:57 Episode 50, steps 118, memory 128196608
...
```
and it's still running stable at least after 100 episodes.

Here is my environment
```
(opensim-rl) lukasz@lukasz-MOBILIZE-NMBL:~$ conda list
# packages in environment at /home/lukasz/anaconda2/envs/opensim-rl:
#
certifi                   2017.4.17                 <pip>
chardet                   3.0.4                     <pip>
freeglut                  3.0.0                         4    kidzik
funcsigs                  1.0.2                    py27_0  
gym                       0.9.2                     <pip>
h5py                      2.7.0               np113py27_0  
hdf5                      1.8.17                        1  
idna                      2.5                       <pip>
keras                     2.0.2                    py27_0  
keras-rl                  0.3.0                     <pip>
libgcc                    5.2.0                         0  
libgfortran               3.0.0                         1  
libgpuarray               0.6.4                         0  
libprotobuf               3.2.0                         0  
mako                      1.0.6                    py27_0  
markupsafe                0.23                     py27_2  
mkl                       2017.0.1                      0  
mkl-service               1.1.2                    py27_3  
mock                      2.0.0                    py27_0  
nose                      1.3.7                    py27_1  
numpy                     1.13.0                   py27_0  
numpy                     1.13.1                    <pip>
openblas                  0.2.19                        0    kidzik
opensim                   4.0.0                   py27_12    kidzik
openssl                   1.0.2l                        0  
osim-rl                   1.4.1                     <pip>
pbr                       1.10.0                   py27_0  
pip                       9.0.1                    py27_1  
protobuf                  3.2.0                    py27_0  
psutil                    5.2.2                     <pip>
pyglet                    1.2.4                     <pip>
pygpu                     0.6.4                    py27_1  
python                    2.7.13                        0  
pyyaml                    3.12                     py27_0  
readline                  6.2                           2  
requests                  2.18.2                    <pip>
scipy                     0.19.0              np113py27_0  
setuptools                27.2.0                   py27_0  
six                       1.10.0                   py27_0  
six                       1.10.0                    <pip>
sqlite                    3.13.0                        0  
tabulate                  0.7.7                     <pip>
tensorflow                1.1.0                    py27_0    conda-forge
theano                    0.9.0                    py27_0  
tk                        8.5.18                        0  
urllib3                   1.22                      <pip>
werkzeug                  0.12.2                   py27_0  
wheel                     0.29.0                   py27_0  
yaml                      0.1.6                         0  
zlib                      1.2.8                         3  
```
Ubuntu 16.04 and my kernel is 4.4.0-72-generic.

It clearly not a proof that things are ok -- it just shows that it'll be harder to fix... I will try to reproduce it on AWS.

**Comment by AdamStelmaszczyk** [comment]
That sheds some light.

So, with Python 2 and 3, kernel 4.8.0-53, there is memory leak.

But, with Python 2 and kernel 4.4.0-72, there isn't.

No idea if kernel version is relevant, there could be some other difference between these envs that matters.

If you are not experiencing leaks, then the leaks in C++ code of OpenSim apparently aren't on the execution path.

Also, I think we may exclude [SWIG](https://www.google.pl/search?q=swig+memory+leak&oq=swig+me&aqs=chrome.0.69i59j69i60j0j69i57j0l2.1358j0j7&sourceid=chrome&ie=UTF-8) as culprit, because opensim from conda repo that we both used was built with the same version of SWIG.

So the open question is, what's the source of memleak. Current hypotheses: kernel version, something else.

**Comment by AdamStelmaszczyk** [comment]
I checked on AWS, it just takes 5 minutes with prepared AMI (Python 2, kernel 4.4.0), following these [instructions](https://github.com/stanfordnmbl/osim-rl/blob/master/docs/AMI.md), plus `conda install psutil` and copying `memleak.py`.

No memory leak.

**Comment by AdamStelmaszczyk** [comment]
On one server with kernel 4.9.30 there is memory leak with Python 3.6, but not with Python 2.7.

Maybe new OpenSim bindings for Python 3.6 are leaking.

BTW in the conda channel `kidzik` there is older version of opensim, for Python3 it will be nice to put there newer version, because it has [this](https://github.com/opensim-org/opensim-core/pull/1862) helpful patch. Otherwise one have to manually tweak `setup.py` file.

**Comment by AdamStelmaszczyk** [comment]
It's easy to reproduce memory leak with Python 3.6 on AWS with your AMI.

Follow these [instructions](https://github.com/stanfordnmbl/osim-rl/blob/master/docs/AMI.md), then:

```
conda create -n py3 -c kidzik opensim git tensorflow psutil python=3.6
pip install git+https://github.com/stanfordnmbl/osim-rl.git
```

Copy and run [`memleak.py`](https://github.com/stanfordnmbl/osim-rl/issues/10#issuecomment-322290035).

**Comment by kidzik** [comment]
One update on my side in Python 2.7:
`10:55:07 Episode 14451, steps 116, memory 128729088`
so there might be a leak here as well but rather marginal.

Indeed, in Python 3.6 I get the leak. This would indicate that the actual problem might be in OpenSim itself and not in this package. For the challenge, we can't guarantee the full support for Python 3.6 since it's very recent in OpenSim. But I will further investigate it with the team. Thanks for all the updates!

**Comment by AdamStelmaszczyk** [comment]
I compiled and installed Python 2.7 OpenSim bindings following [this](https://github.com/opensim-org/opensim-core#ubuntu-1604-xenial-xerus-and-ubuntu-1610-yakkety-yak). With SWIG 3.0.10 and g++ 6.3.0.

Now it's leaking.

With `opensim` from `kidzik` channel there's no leak.

With what command did you compile `opensim` that is in `kidzik` channel (for Python 2)? What g++ and SWIG version?

**Comment by chrisdembia** [comment]
> Now it's leaking.

This supports the idea that the memory leak was introduced in commits to `opensim-core` between when the python 2.7 and the python 3 opensim packages were created (based on discussion with @kidzik).

**Comment by AdamStelmaszczyk** [comment]
Good idea, might be.

@kidzik Do you know on what commit `opensim-core` was when built and stored in channel `kidzik` (for Python 2)?

**Comment by kidzik** [comment]
Sure, it's compiled with this conda receipe https://github.com/opensim-org/conda-opensim/blob/master/opensim/meta.yaml, which means it's this commit in opensim-core
`399c8d57a779dd5dde2916192f8b92bfc959e269`. I don't know which GCC was used though.

**Comment by kidzik** [comment]
Sure, it's compiled with this conda receipe https://github.com/opensim-org/conda-opensim/blob/master/opensim/meta.yaml, which means it's this commit in opensim-core
`399c8d57a779dd5dde2916192f8b92bfc959e269`. I don't know which GCC was used though.

**Comment by AdamStelmaszczyk** [comment]
Great, with https://github.com/opensim-org/opensim-core/commit/399c8d57a779dd5dde2916192f8b92bfc959e269 there's no leak with Python 2.

To find which commit introduced it, [`git bisect`](https://stackoverflow.com/questions/4713088/how-to-use-git-bisect) is ideal:

```
(py2) astelma@chuck:~/opensim-core$ git bisect start
(py2) astelma@chuck:~/opensim-core$ git bisect bad
(py2) astelma@chuck:~/opensim-core$ git bisect good 399c8d57a779dd5dde2916192f8b92bfc959e269
Bisecting: 306 revisions left to test after this (roughly 8 steps)
[3c12ca81228b93c8198aa44d09803bce5e8aa021] Merge branch 'master' into tgcs2017
```

There are 306 commits to check, however with binary search one needs 8 steps or less.

**Comment by kidzik** [comment]
Good point, thanks, Adam! With @chrisdembia we were thinking that for the challenge we may just stick to https://github.com/opensim-org/opensim-core/commit/399c8d57a779dd5dde2916192f8b92bfc959e269 and cherry pick python3 commits -- this should help us avoid the memory leaks and make the python3 and python2 environments more comparable.

**Comment by kidzik** [comment]
Now the new fixed builds for python 3.6 are available on all platforms, so this error shouldn't happen any more.

**Comment by mitkof6** [comment]
This should be fixed in new versions of opensim-core.

https://github.com/opensim-org/opensim-core/pull/2885
